packages/jellyfish-api-core/__tests__/category/icxorderbook/listHTLCs.test.ts

```diff
@@ -0,0 +1,409 @@
+
+    return makeOfferTxId
+  }
+
+  it('Should list htlcs for particular offer', async () => {
+    // create order - maker
+    const order: ICXOrder = {
+      tokenFrom: idDFI,
+      chainTo: 'BTC',
+      ownerAddress: accountDFI,
+      receivePubkey: '037f9563f30c609b19fd435a19b8bde7d6db703012ba1aba72e9f42a87366d1941',
+      amountFrom: new BigNumber(15),
+      orderPrice: new BigNumber(0.01)
+    }
+    let result: ICXGenericResult = await client.icxorderbook.createOrder(order, [])
+    const createOrderTxId = result.txid
+    await container.generate(1)
+
+    // list ICX orders anc check
+    let orders: Record<string, ICXOrderInfo| ICXOfferInfo> = await client.icxorderbook.listOrders()
+    await checkDFISellOrderDetails(container, order, createOrderTxId, orders as Record<string, ICXOrderInfo>)
+
+    // make offer to partial amout 10 DFI - taker
+    const offer: ICXOffer = {
+      orderTx: createOrderTxId,
+      amount: new BigNumber(0.10), // 0.10 BTC = 10 DFI
+      ownerAddress: accountBTC
+    }
+    const accountBTCBeforeOffer = await container.call('getaccount', [accountBTC, {}, true])
+    result = await client.icxorderbook.makeOffer(offer, [])
+    const makeOfferTxId = result.txid
+    await container.generate(1)
+
+    const accountBTCAfterOffer = await container.call('getaccount', [accountBTC, {}, true])
+
+    // check fee of 0.01 DFI has been reduced from the accountBTCBeforeOffer[idDFI]
+    // Fee = takerFeePerBTC(inBTC) * amount(inBTC) * DEX DFI per BTC rate
+    expect(Number(accountBTCAfterOffer[idDFI])).toStrictEqual(Number(accountBTCBeforeOffer[idDFI]) - Number(0.01))
+
+    // List the ICX offers for orderTx = createOrderTxId and check
+    orders = await client.icxorderbook.listOrders({ orderTx: createOrderTxId })
+    expect(Object.keys(orders).length).toBe(2) // extra entry for the warning text returned by the RPC atm.
+    await checkDFIBuyOfferDetails(container, offer, makeOfferTxId, orders as Record<string, ICXOfferInfo>)
+
+    const accountDFIBeforeDFCHTLC = await container.call('getaccount', [accountDFI, {}, true])
+
+    // create DFCHTLC - maker
+    const DFCHTLC: HTLC = {
+      offerTx: makeOfferTxId,
+      amount: new BigNumber(10), // in  DFC
+      hash: '957fc0fd643f605b2938e0631a61529fd70bd35b2162a21d978c41e5241a5220',
+      timeout: 500
+    }
+    const DFCHTLCTxId = (await client.icxorderbook.submitDFCHTLC(DFCHTLC)).txid
+    await container.generate(1)
+
+    const accountDFIAfterDFCHTLC = await container.call('getaccount', [accountDFI, {}, true])
+    expect(Number(accountDFIAfterDFCHTLC[idDFI])).toStrictEqual(Number(accountDFIBeforeDFCHTLC[idDFI]) - Number(0.01))
+
+    // List htlc anc check
+    let listHTLCOptions: ICXListHTLCOptions = {
+      offerTx: makeOfferTxId
+    }
+    let HTLCs: Record<string, ICXDFCHTLCInfo| ICXEXTHTLCInfo| ICXClaimDFCHTLCInfo> = await client.icxorderbook.listHTLCs(listHTLCOptions)
```

Comment:
Here the client.icxorderbook.listHTLCs() function can return 3 types of structures depending on the input. I used a union for this. But on the client side code, returned structure has to be narrowed down depends on either the input or certain properties of the output structure. Here it is narrowed down based on HTLCs[DFCHTLCTxId].type (inside checkDFCHTLCDetails())
If anybody has better idea to deal this kind of situations please comment.